test/distributed/_tensor/experimental/test_local_map.py

```diff
@@ -0,0 +1,200 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates
+# Owner(s): ["oncall: distributed"]
+
+import torch
+import torch.distributed._functional_collectives as funcol
+from torch.distributed._tensor import (
+    distribute_tensor,
+    init_device_mesh,
+    Replicate,
+    Shard,
+)
+from torch.distributed._tensor.debug import CommDebugMode
+from torch.distributed._tensor.experimental import local_map
+from torch.testing._internal.common_utils import run_tests
+from torch.testing._internal.distributed._tensor.common_dtensor import (
+    DTensorTestBase,
+    with_comms,
+)
+
+DEVICE_TYPE = "cpu"
+
+
+def mm_forward(device_mesh, W, X):
+    return torch.mm(W, X)
+
+
+def mm_allreduce_forward(device_mesh, W, X):
+    partial_sum_tensor = torch.mm(W, X)
+    reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+    return reduced_tensor
```

Comment:
nit: maybe explicitly wait for this collective call, otherwise this would leak AsyncColelctiveTensor out